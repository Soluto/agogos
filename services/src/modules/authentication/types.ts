import { ClientOptions } from 'jwks-rsa';

export interface IssuerConfig {
  authority: string;
  jwksConfig?: Partial<ClientOptions>;
  audience?: string | string[];
  authenticatedPaths: string[];
  description?: string;
}

export interface AuthenticationConfig {
  jwt?: Record<string, IssuerConfig>;
  anonymous?: {
    publicPaths: string[];
    rejectAuthorizationHeader?: boolean;
  };
}
