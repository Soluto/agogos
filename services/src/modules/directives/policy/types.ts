import {PolicyArgsObject, PolicyAttachments} from '../../resource-repository/types';

export type Policy = {
    namespace: string;
    name: string;
    args?: PolicyArgsObject;
};

// args here contain the final values after param injection
export type PolicyExecutionContext = {
    namespace: string;
    name: string;
    policyAttachments: PolicyAttachments;
    args?: PolicyArgsObject;
    query?: QueryResults;
};

export type QueryResults = {
    [name: string]: any;
};

export type PolicyExecutionResult = {
    done: boolean;
    allow?: boolean;
    query?: {
        type: string;
        code: string;
    };
};

export type GraphQLArguments = {
    [name: string]: any;
};
