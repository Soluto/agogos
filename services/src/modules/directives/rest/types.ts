export interface KeyValue {
  key: string;
  value: string;
}

export interface RestParams {
  url: string;
  method?: string;
  body?: string;
  bodyArg?: string;
  query?: KeyValue[];
  headers?: KeyValue[];
  timeoutMs?: number;
}
