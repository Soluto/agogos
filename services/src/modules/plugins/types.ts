import { BaseSchema } from '../base-schema';
import { ResourceGroup } from '../resource-repository';

export type ValueOrPromise<T> = T | Promise<T>;

export interface StitchPlugin {
  name: string;

  addArgumentInjectionGlobals?(): ValueOrPromise<Record<string, unknown>>;

  transformResourceGroup?(resourceGroup: ResourceGroup): ValueOrPromise<ResourceGroup>;

  transformBaseSchema?(baseSchema: BaseSchema): ValueOrPromise<BaseSchema>;
}
