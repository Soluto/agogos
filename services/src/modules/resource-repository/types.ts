import { RemoteSchema } from '../directives/gql';
import { Policy, LoadedPolicy } from '../directives/policy/types';

export type PolicyAttachments = Record<string, LoadedPolicy>;
export type PolicyQueryVariables = Record<string, unknown>;
export type PolicyArgsObject = Record<string, unknown>;

export interface ResourceGroup {
  schemas: Schema[];
  upstreams: Upstream[];
  upstreamClientCredentials: UpstreamClientCredentials[];
  policies: PolicyDefinition[];
  // policyAttachments are compiled from the Rego code in opa policies, they are not directly modified by users
  policyAttachments?: PolicyAttachments;
  basePolicy?: Policy;
  remoteSchemas?: RemoteSchema[];
}

export interface Resource {
  metadata: ResourceMetadata;
}

export interface ResourceMetadata {
  namespace: string;
  name: string;
}

export interface Schema extends Resource {
  schema: string;
}

export interface Upstream extends Resource {
  host: string;
  auth: {
    type: AuthType;
    activeDirectory: {
      authority: string;
      resource: string;
    };
  };
}

export interface UpstreamClientCredentials extends Resource {
  authType: AuthType;
  activeDirectory: {
    authority: string;
    clientId: string;
    clientSecret: string;
  };
}

export interface PolicyDefinition extends Resource {
  type: PolicyType;
  code: string;
  args?: PolicyArgsObject;
  query?: PolicyQuery;
  shouldOverrideBasePolicy?: boolean;
}

export interface PolicyQuery {
  gql: string;
  variables?: PolicyQueryVariables;
}

export interface FetchLatestResult {
  isNew: boolean;
  resourceGroup: ResourceGroup;
}

export type UpdateOptions = { registry: boolean };

export interface IResourceRepository {
  fetchLatest(): Promise<FetchLatestResult>;
  update(rg: ResourceGroup, options?: UpdateOptions): Promise<void>;
  writePolicyAttachment(filename: string, content: Buffer): Promise<void>;
}

enum AuthType {
  ActiveDirectory = 'ActiveDirectory',
}

export enum PolicyType {
  opa = 'opa',
}
