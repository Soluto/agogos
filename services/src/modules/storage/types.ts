export interface Storage {
  fileStats(filePath: string): Promise<FileStats>;
  writeFile(filePath: string, content: string | Buffer): Promise<void>;
  listFiles(folderPath: string): Promise<listFilesItem[]>;

  readFile(filePath: string): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: false }): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: true }): Promise<{ content: string }>;
  readFile(filePath: string, options?: { asString?: boolean }): Promise<{ content: Buffer | string }>;
}

export type FileStats = {
  lastModified: Date;
};

export type listFilesItem = {
  filePath: string;
  lastModified: Date;
};
