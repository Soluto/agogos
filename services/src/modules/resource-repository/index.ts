export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
