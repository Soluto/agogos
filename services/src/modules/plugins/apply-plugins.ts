import { promises as fs } from 'fs';
import { join } from 'path';
import { BaseSchema } from '../base-schema';
import { pluginsDir, pluginsConfig } from '../config';
import logger from '../logger';
import { ResourceGroup } from '../resource-repository';
import { StitchPlugin } from './types';

export const plugins: StitchPlugin[] = [];
export let argumentInjectionGlobals: Record<string, unknown> = {};

export async function loadPlugins() {
  if (!pluginsDir) return;
  const pluginList = await fs.readdir(pluginsDir);

  await Promise.all(
    pluginList.map(async name => {
      try {
        const pluginImport = await import(join(pluginsDir!, name));
        const plugin: Partial<Omit<StitchPlugin, 'name'>> =
          typeof pluginImport === 'function' ? await pluginImport.call() : pluginImport;

        if (plugin.configure) {
          await plugin.configure(pluginsConfig?.[name]);
        }

        plugins.push({ name, ...plugin });
        logger.info({ name }, `Plugin ${name} has been loaded successfully.`);
      } catch (err) {
        logger.error({ err, name }, 'Failed to load plugin');
        throw err;
      }
    })
  );

  try {
    argumentInjectionGlobals = await buildArgumentInjectionGlobals();
  } catch (err) {
    logger.error({ err }, 'Failed to build argument injection globals');
    throw err;
  }
}

export async function buildArgumentInjectionGlobals(): Promise<Record<string, unknown>> {
  const globals = await plugins
    .filter(p => p.addArgumentInjectionGlobals)
    .reduce(async (resultPromise, curPlugin) => {
      try {
        logger.trace(
          { plugin: curPlugin.name, method: 'addArgumentInjectionGlobals' },
          `Applying addArgumentInjectionGlobals of ${curPlugin.name} plugin...`
        );
        const resultGlobals = await resultPromise;
        const curPluginGlobals = await curPlugin.addArgumentInjectionGlobals!();

        logger.trace(
          { plugin: curPlugin.name, method: 'addArgumentInjectionGlobals' },
          `addArgumentInjectionGlobals of ${curPlugin.name} plugin was applied successfully.`
        );

        return { ...resultGlobals, ...curPluginGlobals };
      } catch (err) {
        const message = `Plugin "${curPlugin.name}" failed to execute "addArgumentInjectionGlobals" with error: ${err}`;
        logger.warn({ err, plugin: curPlugin.name, method: 'addArgumentInjectionGlobals' }, message);
        throw new Error(message);
      }
    }, Promise.resolve({}));
  return globals;
}

export async function transformResourceGroup(resourceGroup: ResourceGroup): Promise<ResourceGroup> {
  let rg: ResourceGroup = resourceGroup;
  const pluginsToApply = plugins.filter(p => p.transformResourceGroup);
  for (const curPlugin of pluginsToApply) {
    try {
      logger.trace(
        { plugin: curPlugin.name, method: 'transformResourceGroup' },
        `Applying transformResourceGroup of ${curPlugin.name} plugin...`
      );
      rg = await curPlugin.transformResourceGroup!(rg);
      logger.trace(
        { plugin: curPlugin.name, method: 'transformResourceGroup' },
        `transformResourceGroup of ${curPlugin.name} plugin was applied successfully.`
      );
    } catch (err) {
      const message = `Plugin "${curPlugin.name}" failed to execute "transformResourceGroup" with error: ${err}`;
      logger.warn({ err, plugin: curPlugin.name, method: 'transformResourceGroup' }, message);
      throw new Error(message);
    }
  }
  return rg;
}

export async function transformBaseSchema(baseSchema: BaseSchema): Promise<BaseSchema> {
  let bs: BaseSchema = baseSchema;
  const pluginsToApply = plugins.filter(p => p.transformBaseSchema);
  for (const curPlugin of pluginsToApply) {
    try {
      logger.trace(
        { plugin: curPlugin.name, method: 'transformBaseSchema' },
        `Applying transformBaseSchema of ${curPlugin.name} plugin...`
      );
      bs = await curPlugin.transformBaseSchema!(bs);
      logger.trace(
        { plugin: curPlugin.name, method: 'transformBaseSchema' },
        `transformBaseSchema of ${curPlugin.name} plugin was applied successfully.`
      );
    } catch (err) {
      const message = `Plugin "${curPlugin.name}" failed to execute "transformBaseSchema" with error: ${err}`;
      logger.warn({ err, plugin: curPlugin.name, method: 'transformBaseSchema' }, message);
      throw new Error(message);
    }
  }
  return bs;
}
