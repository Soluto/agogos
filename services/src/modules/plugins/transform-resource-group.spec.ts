import * as _ from 'lodash';
import { PolicyType, ResourceGroup } from '../resource-repository';
import { plugins, transformResourceGroup } from './apply-plugins';
import { StitchPlugin } from './types';

const resourceGroup: ResourceGroup = {
  schemas: [
    {
      metadata: {
        namespace: 'ns',
        name: 'old-name',
      },
      schema: `
              type Query {
                foo: String!
              }
            `,
    },
  ],
  upstreams: [],
  upstreamClientCredentials: [],
  policies: [],
};
const loadedPlugins: StitchPlugin[] = [
  {
    name: 'rename-name',
    version: 'N/A',
    transformResourceGroup(rg: ResourceGroup) {
      return _.set(rg, 'schemas[0].metadata.name', 'new-name');
    },
  },
  {
    name: 'rename-namespace',
    version: 'N/A',
    transformResourceGroup(rg: ResourceGroup) {
      return _.set(rg, 'schemas[0].metadata.namespace', 'new-ns');
    },
  },
  {
    name: 'do-nothing',
    version: 'N/A',
  },
  {
    name: 'transform-resource-group',
    version: 'N/A',
    transformResourceGroup(resourceGroup: ResourceGroup) {
      return {
        ...resourceGroup,
        policies: [
          {
            metadata: {
              namespace: 'ns',
              name: 'policy',
            },
            type: PolicyType.opa,
            code: 'REGO code',
          },
        ],
      };
    },
  },
];

const loadedPluginsWithCrash: StitchPlugin[] = [
  {
    name: 'do-nothing',
    version: 'N/A',
  },
  {
    name: 'throw error',
    version: 'N/A',
    transformResourceGroup() {
      throw new Error('Something bad happened');
    },
  },
];

describe('Plugins: transformResourceGroup', () => {
  test('transformResourceGroup', async () => {
    plugins.splice(0, plugins.length);
    plugins.push(...loadedPlugins);
    const result = await transformResourceGroup(resourceGroup);
    expect(result).toMatchSnapshot();
  });

  test('transformResourceGroup crashing', async () => {
    plugins.splice(0, plugins.length);
    plugins.push(...loadedPluginsWithCrash);
    const result = await transformResourceGroup(resourceGroup).catch(e => e);
    expect(result).toMatchSnapshot();
  });
});
