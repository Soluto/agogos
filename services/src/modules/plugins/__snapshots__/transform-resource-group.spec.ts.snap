// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugins: transformResourceGroup transformResourceGroup 1`] = `
Object {
  "policies": Array [
    Object {
      "code": "REGO code",
      "metadata": Object {
        "name": "policy",
        "namespace": "ns",
      },
      "type": "opa",
    },
  ],
  "schemas": Array [
    Object {
      "metadata": Object {
        "name": "new-name",
        "namespace": "new-ns",
      },
      "schema": "
              type Query {
                foo: String!
              }
            ",
    },
  ],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Plugins: transformResourceGroup transformResourceGroup crashing 1`] = `[Error: Plugin "throw error" failed to execute "transformResourceGroup" with error: Error: Something bad happened]`;
